feat(minifier): compress `(a = _a) != null ? a : b` and `(a = _a) != null ? a.b() : undefined`

sapphi-red opened the pull request:
Compresses `(a = _a) != null ? a : b` to `(a = _a) ?? b` and `(a = _a) != null ? a.b() : undefined` to `(a = _a).b()`. This commonly happens when lowered.

sapphi-red commented:
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/8823?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#8823** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/8823?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/8823?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#8822** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/8822?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* `main`
<hr />

## How to use the Graphite Merge Queue

Add either label to this PR to merge it via the merge queue:
- <kbd>0-merge</kbd> - adds this PR to the back of the merge queue
- <kbd>hotfix</kbd> - for urgent hot fixes, skip the queue and merge this PR next

You must have a Graphite account in order to use the merge queue. Sign up using [this link](https://app.graphite.dev/invite/github/oxc-project?ref=merge-queue-instructions-in-stack-comment&prId=8823).

<h2></h2>
<sub>

An organization admin has enabled the [Graphite Merge Queue](https://graphite.dev/docs/graphite-merge-queue) in this repository.

</sub>

<sub>

Please do not merge from GitHub as this will restart CI on PRs being processed by the merge queue.

</sub>

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

Boshen commented on crates/oxc_minifier/src/peephole/minimize_conditions.rs:
```diff
@@ -698,64 +716,76 @@ impl<'a> PeepholeOptimizations {
         ctx: Ctx<'a, '_>,
     ) -> bool {
-        fn inject(target_expr: &Expression<'_>, expr: &mut Expression<'_>) -> bool {
+        fn inject<'aa>(
```
`'aa` 🤨 I try to avoid inlining functions inside functions.

sapphi-red replied:
The reason why this function is placed inside this function is because this function doesn't make sense without the outer function. I think we can avoid this lifetime parameter by three ways.
1. put this inner function directly in `impl<'a> PeepholeOptimizations` block as `inject_optional_chaining_if_matched_inner`
1. convert this inner function to a closure
1. convert the recursive call into a loop

sapphi-red replied:
Which one should I do? or are there any better ways to do?

Boshen replied:
I prefer `inject_optional_chaining_if_matched_inner`

Boshen commented:
### Merge activity

* **Feb 1, 8:08 PM EST**: The merge label '0-merge' was detected. This PR will be added to the [Graphite merge queue](https://app.graphite.dev/merges?org=oxc-project&repo=oxc) once it meets the requirements.

sapphi-red committed:
feat(minifier): compress `(a = _a) != null ? a : b` and `(a = _a) != null ? a.b() : undefined` (#8823)

Compresses `(a = _a) != null ? a : b` to `(a = _a) ?? b` and `(a = _a) != null ? a.b() : undefined` to `(a = _a).b()`. This commonly happens when lowered.
